src/thor/trace_route_action.cc

```diff
@@ -386,58 +388,62 @@ thor_worker_t::map_match(Api& request, uint32_t best_paths) {
     else {
-      // we make a new route to hold the result of each iteraton of top k
+
       auto& route = *request.mutable_trip()->mutable_routes()->Add();
-      auto origin = match_results.begin();
-      auto destination = match_results.begin();
-
-      size_t last_index = 0;
-      // for each edge in the path
-      for (size_t i = 0; i < path_edges.size(); ++i) {
-        const auto& path_edge = path_edges[i];
-        // while we still have trace points that are on this current edge of the path
-        while (destination != match_results.end() && path_edge.edgeid == destination->edgeid) {
-          // if it's not the first trace point, and it's a break, make a trip path leg
-          if (origin != destination && options.shape(destination - match_results.begin()).type() ==
-                                           valhalla::Location::kBreak) {
-            // turn the origin and locations into real ones with a path edge on them
-            // its possible that meili has a path edge for this one but its also possible
-            // that it doesnt because its interpolated
-            auto* o_loc = options.mutable_shape(origin - match_results.begin());
-            auto* d_loc = options.mutable_shape(destination - match_results.begin());
-            add_path_edge(o_loc, *origin);
-            add_path_edge(d_loc, *destination);
-            // get a slice of the paths edges
-            std::vector<PathInfo> sub_path_edges(path_edges.begin() + last_index,
-                                                 path_edges.begin() + i + 1);
-            // build the leg
-            TripLegBuilder::Build(controller, matcher->graphreader(), mode_costing,
-                                  path_edges.begin() + last_index, path_edges.begin() + i + 1, *o_loc,
-                                  *d_loc, std::list<valhalla::Location>{},
-                                  *route.mutable_legs()->Add(), interrupt, &route_discontinuities);
-            // beginning of next leg will be the end of this leg
-            origin = destination;
-            // store the starting index of the path_edges
-            last_index = i;
+      auto origin_iter = match_results.begin();
+
+      /*the following logic builds leges on edge candidates that generated by meili*/
+
+      // To build legs, the first valid matched location should be on the first edge candidate
+      while (origin_iter->edgeid != m_temp_path_edges.front().edgeid) {
+        ++origin_iter;
+      }
+
+      // The following logic put break points (matches results) on edge candidates to form legs
+      // logic assumes the both match results and edge candidates are topologically sorted in correct
+      // order
+      int last_leg_index = 0;
+      for (int i = 0, n = static_cast<int>(m_temp_path_edges.size()); i < n; ++i) {
+        const auto& path_edge = m_temp_path_edges[i];
+
+        for (auto iter = origin_iter + 1; iter != match_results.end(); ++iter) {
```

Comment:
this makes sense but what happens if there is a discontinuity in the match. also what about the end of the match. similar to when you iterate the origin_iter to find the first valid edge id, what do we do at the end to make sure we end the leg at the rigth place in the case that offline match couldnt match some points? maybe this will be handled in a separate pr, but im wondering what happens now when we encounter those cases

Reply:
There are cases has discontinuity in the match, e.g, two edge candidates are not connected, the logic here in the inner for loop will jump through the matched locations that doesn't have the same edgeid with the edge candidates. As soon as there is a match in edge_ids, we update the origin_iter to the correct place. This mechanism is suppose to handle the discontinuity.

Reply:
If there is an end edge and discontinuity in-between, the inner loop is suppose to update the origin_iter to the matched locations that has the same edgeid with the end edge. I think it handles that case too, correct me if I am wrong

Reply:
If remove the disconnect edge check in line 178 and let the above logic handles the situation, it looks fine for the test cases that I am playing with. However, we need to run some more test to verify more.

Reply:
yeah i guess i just wonder what is the final output if you are supposed to have a leg but there is a discontinuity in the middle of it. you cant or at least you shouldnt call TripLegBuilder::Build for a leg that has a discontinuity it in.

for example what if your input points are:

`break through break through break`

and the match is

`matched matched matched failed matched`

what legs do you return, the second leg has a disconnected path. do you just return one leg?